design/authenticator-client-for-container-platforms.md

```diff
@@ -0,0 +1,211 @@
+  enterprise secrets physically available (even if encrypted at rest) inside a container platform environment.
+- Provide strong authentication via multiple factors ("secret-zero"). At least one of these factors must use a
+  different communication channel from the others.
+- Support a time-limited access token or other credential which will quickly expire and can be frequently refreshed.
+- Protect access tokens against use in unauthorized locations (e.g. stolen and used elsewhere in the network).
+
+And it should:
+
+- Have an implementation as similar as possible across the supported platforms.
+
+## Architecture
+                                        +------------+
+                                        |E. Security |
+                          +------------->   Service  |
+                          |             |            |
+                          |             +------^-----+
+                          |                    |
+                          |                    |
++----------------------------------------------------------+
+|                         |                    |           |
+|                         |                    |           |
+| +------------+   +------+-------+   +--------+-------+   |
+| |B. Container|   |C. Application+-->+D. Authenticator|   |
+| |   Platform +-->+   Containers |   +----------------+   |
+| |   Engine   |   +--------------+                        |
+| +------------+                                           |
+|                                                          |
+|                                   A: Compute Cluster     |
++----------------------------------------------------------+
+
+A. Compute Cluster - A pool of servers or VMs which run a container engine.
+B. Container Platform Engine - Schedules application containers across the Compute Cluster.
+C. Application Containers - Individual instances of application code which are scheduled onto the Compute Cluster
+   by the Platform Engine
+D. Authenticator - A platform application whose function is to verify the identity of Application Containers and issue
+   them identity credentials. Runs in the compute cluster.
+E. Security Service - A service which provides security methods such as role-based access control, permission checks,
+   auditing, and access to secrets. May run inside or outside the compute cluster.
+
+## General design
+
+The Authenticator provides two operations: `login` and `authenticate`. `login` is performed once by each client to
+obtain a signed certificate. `authenticate` is performed many times by each client in order to obtain a short-lived
+access token. Each access token thus obtained is encrypted by the client's public key.
+
+The `login` method resists identity spoofing by installing the signed certificate onto the client using a Container
+Platform Engine API.
+
+Once the client has obtained its signed certificate, it uses the certificate to authenticate, thereby  obtaining a
```

Comment:
The server injects the client certificate before this part is done which should be mentioned.